src/libs/sys-util-rs/src/mount.rs

```diff
@@ -0,0 +1,808 @@
+            if e.kind() != std::io::ErrorKind::AlreadyExists {
+                return Err(e.into());
+            }
+        }
+    } else {
+        fs::OpenOptions::new()
+            .mode(MOUNT_PERM)
+            .write(true)
+            .create(true)
+            .open(dst)?;
+    }
+
+    Ok(())
+}
+
+/// Create a mountpoint on Linux.
+pub fn linux_mount(target: &str, m: &Mount) -> Result<()> {
+    fail::fail_point!("linux_mount", |_| {
+        Err(anyhow!("linux mount fail point injection"))
+    });
+
+    let f_type = &m.fs_type;
+    // Follow the same algorithm as Containerd: reserve 512 bytes to avoid hitting one page limit
+    // of mounting argument buffer.
+    let (chdir, options) = if f_type == "overlay" && option_size(&m.options) >= PAGESIZE - 512 {
+        info!(
+            sl!(),
+            "overlay mount option too long, maybe failed to mount"
+        );
+        compact_lowerdir_option(&m.options)
+    } else {
+        ("".to_string(), m.options.clone())
+    };
+
+    let m_opts = parse_mount_option(&options)?;
+    let flag = m_opts.flag;
+    let propagation_types =
+        MsFlags::MS_SHARED | MsFlags::MS_PRIVATE | MsFlags::MS_SLAVE | MsFlags::MS_UNBINDABLE;
+    let o_flag = flag & (!propagation_types);
+    let data = m_opts.data.as_str();
+
+    // normal mount or remount to change fs-specific options
+    if (flag & MsFlags::MS_REMOUNT) == MsFlags::empty() || !data.is_empty() {
+        return mount_at(
+            chdir.as_str(),
+            m.source.as_str(),
+            target,
+            f_type.as_str(),
+            o_flag,
+            data,
+        );
+    }
+
+    // remount to change propagation
+    if (flag & propagation_types) != MsFlags::empty() {
+        let propagation_flag = propagation_types | MsFlags::MS_REC | MsFlags::MS_SILENT;
+        return mount(
+            Some(""),
+            target,
+            Some(""),
+            flag & propagation_flag,
+            Some(""),
+        )
+        .map_err(|e| Error::Mount("".to_string(), target.to_string(), e));
+    }
+
+    let bro_flag = MsFlags::MS_BIND | MsFlags::MS_RDONLY;
+    if (o_flag & bro_flag) == bro_flag {
+        return mount(
+            Some(""),
+            target,
+            Some(""),
+            o_flag & MsFlags::MS_REMOUNT,
+            Some(""),
+        )
+        .map_err(|e| Error::BindMount("".to_string(), target.to_string(), e));
+    }
+
+    // the rest of remounts, e.g. remount from ro back to rw
+    mount(Some(""), target, Some(""), flag, Some(""))
+        .map_err(|e| Error::Mount("".to_string(), target.to_string(), e))
+}
+
+fn option_size(opts: &[String]) -> usize {
+    opts.iter().fold(0, |acc, x| acc + x.len() + 1)
+}
+
+fn parse_mount_option(options: &[String]) -> Result<MountOption> {
+    let mut flag: MsFlags = MsFlags::empty();
+    let mut data: Vec<String> = Vec::new();
+
+    let mut clear_flags: HashMap<&str, MsFlags> = HashMap::new();
```

Comment:
It's convenient to use HashMap::from_iter() with latest rustc as
```
let v: HashMap<_, _> = HashMap::from_iter([(1,2), (3,4)]);
```
But it doesn't work well with some older rustc, it needs a `vec![]` to work:
```
let v: HashMap<_, _> = HashMap::from_iter(vec![(1,2), (3,4)]);
```

For a HashMap, it doesn't make a big difference between
```
HashMap::new() + HashMap::insert()
```
and
```
HashMap::from_iter()
```
because HashMap::from_iter() calls HashMap::insert() too.

By the way, we could optimize the code in another way by using lazy_static instead.